### PR TITLE
Initial commit for PluggableSASLAuthenticationProvider

### PR DESCRIPTION
**Pull request by bothejjms**

### CODE REVIEW

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,6 +1,5 @@
-#Fri Jun 02 11:09:49 IST 2017
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-2.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.2.1-ms-201712061-bin.zip
```
Comment (andreikeis) [inline gradle/wrapper/gradle-wrapper.properties]: this does not look right. ms-?

File: settings.gradle [hunk 2]
```diff
@@ -1,12 +1,11 @@
...
- * This settings file was auto generated by the Gradle buildInit task
- * by 'rahulj' at '6/2/17 11:09 AM' with Gradle 2.1
+ * This settings file was generated by the Gradle 'init' task.
  *
  * The settings file is used to specify which projects to include in your build.
  * In a single project build this file can be empty or even removed.
  *
  * Detailed information about configuring a multi-project build in Gradle can be found
- * in the user guide at http://gradle.org/docs/2.1/userguide/multi_project_builds.html
+ * in the user guide at https://docs.gradle.org/4.2.1-ms-201712061/userguide/multi_project_builds.html
```
Comment (andreikeis) [inline settings.gradle]: replace with http://gradle.org/docs or remove the comments.

File: settings.gradle [hunk 3]
```diff
@@ -16,4 +15,4 @@ include 'api'
 include 'services:webservice'
 */
 
-rootProject.name = 'zookeeper-ldap-roles'
+rootProject.name = 'zookeeper-athorizers'
```
Comment (andreikeis) [inline settings.gradle]: typo. authorizes.

File: src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+import org.apache.log4j.Logger;
+import org.apache.zookeeper.server.auth.SASLAuthenticationProvider;
+
+import com.ms.zookeeper.authz.ZkAuthorizer;
+
+/**
+ * Extending Sasl authentication provider for ZooKeeper with pluggable authorizers.
+ * 
+ * ZooKeeper uses a plugin mechanism to provide authentication and authorisation
```
Comment (andreikeis) [inline src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java]: typo - authorization.

File: src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+ * This class implements the ZK AuthenticationProvider to act as a server side
+ * Authenticator, but also overloads the constructor and supplies a
+ * getAuthToken() method to be used on the client side to get a valid auth
+ * token. This lets us keep the kerberos related enhancements to ZooKeeper in a
```
Comment (andreikeis) [inline src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java]: I do not think this comment make sense. How is it related to kerberos enhancements? leftover?

File: src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+ * single place. Easier to test compatibility with future ZK versions.
+ * 
+ * Note: The server process must run under a production id with a valid keytab
```
Comment (andreikeis) [inline src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java]: review comments, make them relevant.

File: src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+    private static final Logger LOG = Logger.getLogger(PluggableSASLAuthenticationProvider.class);
+
+    private static final HashMap<String, ZkAuthorizer> AUTHORIZERS;
+    static {
+        AUTHORIZERS = getAuthorizers();
+    }
+
+    private static HashMap<String, ZkAuthorizer> getAuthorizers() {
+        String authzString = System.getProperty("zookeeper.msAuthorizers", "user,file");
```
Comment (andreikeis) [inline src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java]: remove msAuthorizers.

File: src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+    /*
+     * @see
+     * org.apache.zookeeper.server.auth.AuthenticationProvider#isValid(java.
+     * lang.String)
+     */
+    @Override
+    public boolean isValid(String id) {
+        /**
+         * id is the id set by the client as the second part of the scheme+id
```
Comment (andreikeis) [inline src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java]: can you rephrase the comment to be (more) readable?

File: src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        if (AUTHORIZERS.containsKey(authorizer)) {
+            return true;
+        }
+        LOG.info("Invalid authorizer : " + authorizer);
+        return false;
+    }
+
+    /*
+     * id is the user that is trying to access the znode aclExpr is of the form
+     * protocol://data. eg - user://zapptest or ldap://zappcore
```
Comment (andreikeis) [inline src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java]: remove zapptest/zappcore., make comment relevant.

File: src/main/java/com/ms/zookeeper/authz/ZkUserAuthorizer.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    private static  ZkUserAuthorizer instance = null;
+
+    private static final String HOST_PREFIX = "host/";
+
+    private static final String ARBITRARY_HOST = HOST_PREFIX + "*";
+
+    private final Set<String> validRealms;
+
+    public ZkUserAuthorizer() {
+        String realmsString = System.getProperty("zookeeper.validKerberosRealms", "");
```
Comment (andreikeis) [inline src/main/java/com/ms/zookeeper/authz/ZkUserAuthorizer.java]: you keep mentioning kerberos, but it should work with any auth provider, no? In the file, this is just a list of strings that need to match?

File: src/main/java/com/ms/zookeeper/authz/ZkUserAuthorizer.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        }
+        return instance;
+    }
+
+    private boolean isValidRealm(final String realm) {
+        return validRealms.contains(realm.toLowerCase());
+    }
+
+    private boolean allowed(final String reqUser, final String aclUser) {
+        LOG.debug(" Allowing user " + reqUser + " because user matches ACL user " + aclUser);
```
Comment (andreikeis) [inline src/main/java/com/ms/zookeeper/authz/ZkUserAuthorizer.java]: Nit - remove space before "Allowing". I would suggest shorten the message, and use grant/deny rather than allowing/not allowing.

File: src/main/java/com/ms/zookeeper/authz/ZkUserAuthorizer.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    public static ZkUserAuthorizer getInstance() {
+        if (instance == null) {
+            instance = new ZkUserAuthorizer();
+        }
+        return instance;
+    }
+
+    private boolean isValidRealm(final String realm) {
```
Comment (andreikeis) [inline src/main/java/com/ms/zookeeper/authz/ZkUserAuthorizer.java]: why do you care about valid realm? Can't the file just contain full principal, including the realm?

File: src/test/java/com/ms/zookeeper/TestZkUserAuthorizer.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
+package com.ms.zookeeper;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import org.junit.Test;
+
+import com.ms.zookeeper.authz.ZkUserAuthorizer;
```
Comment (andreikeis) [inline src/test/java/com/ms/zookeeper/TestZkUserAuthorizer.java]: com.ms - wrong namespace.

File: src/test/java/com/ms/zookeeper/TestZkUserAuthorizer.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+        ZkUserAuthorizer auth = new ZkUserAuthorizer();
+        assertTrue(auth.authorize("username", "user://username"));
+        assertFalse(auth.authorize("username", "user://otherusername"));
+    }
+
+    @Test
+    public void testUserAuthWithRealms() {
+        System.setProperty("zookeeper.validKerberosRealms", "realm,realm2");
+        ZkUserAuthorizer auth = new ZkUserAuthorizer();
+        assertTrue(auth.authorize("username@REALM", "user://username"));
```
Comment (andreikeis) [inline src/test/java/com/ms/zookeeper/TestZkUserAuthorizer.java]: back to previous q - do we have to strip realm?

File: src/test/java/com/ms/zookeeper/TestZkUserAuthorizer.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        assertFalse(auth.authorize("username@REALM2", "user://otherusername"));
+
+        assertFalse(auth.authorize("username@foo.bar", "user://username"));
+    }
+
+    @Test
+    public void testUserAuthWithHosts() {
+        System.setProperty("zookeeper.validKerberosRealms", "realm,realm2");
+        ZkUserAuthorizer auth = new ZkUserAuthorizer();
+        assertTrue(auth.authorize("host/hostname", "user://host/*"));
```
Comment (andreikeis) [inline src/test/java/com/ms/zookeeper/TestZkUserAuthorizer.java]: same here - can users and host and everything else be treated same way?

File: src/test/resources/log4j.properties [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+# Example with rolling log file
+#log4j.rootLogger=DEBUG, CONSOLE, ROLLINGFILE
+
+# Example with rolling log file and tracing
+#log4j.rootLogger=TRACE, CONSOLE, ROLLINGFILE, TRACEFILE
+
+#
+# Log INFO level and above messages to the console
+#
+log4j.appender.CONSOLE=org.apache.log4j.ConsoleAppender
```
Comment (andreikeis) [inline src/test/resources/log4j.properties:20]: does logging configuration belong to this component? it should be defined for zk, not for authorizer, which is plugin.

**Review by andreikeis — CHANGES_REQUESTED** [review]

File: src/test/resources/log4j.properties (hunk 16 above)
Comment (bothejjms) [reply src/test/resources/log4j.properties:20]: This is for the unit tests only. We need log setting so we can see the logs in stdout when we run the unit tests

File: src/test/java/com/ms/zookeeper/TestZkUserAuthorizer.java (hunk 14 above)
Comment (bothejjms) [reply src/test/java/com/ms/zookeeper/TestZkUserAuthorizer.java]: We do strip the realm internally. I don't want to fork this functionality so it is easier to maintain. 
This is an optional feature. If validKerberosRealms are not configure than strip won't work

File: src/test/java/com/ms/zookeeper/TestZkUserAuthorizer.java (hunk 15 above)
Comment (bothejjms) [reply src/test/java/com/ms/zookeeper/TestZkUserAuthorizer.java]: the extra function here is the wildcard (*) which we don't have for user. Again an internal feature. Treadmill was using it. Maybe already abandoned and we use only the file authorizer.

File: src/test/java/com/ms/zookeeper/TestZkUserAuthorizer.java (hunk 13 above)
Comment (bothejjms) [reply src/test/java/com/ms/zookeeper/TestZkUserAuthorizer.java]: I was wondering about that. Since it is in an ms repo. I thought it makes sense to use ms namespace. What would you prefer?

File: src/main/java/com/ms/zookeeper/authz/ZkUserAuthorizer.java (hunk 12 above)
Comment (bothejjms) [reply src/main/java/com/ms/zookeeper/authz/ZkUserAuthorizer.java]: See my other comment. It is an optional feature which I wanted to maintain in open source as well.

File: src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java (hunk 7 above)
Comment (bothejjms) [reply src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java]: We need a configuration to define which plugins should be loaded. We use different plugins in different deployments

File: gradle/wrapper/gradle-wrapper.properties (hunk 1 above)
Comment (bothejjms) [reply gradle/wrapper/gradle-wrapper.properties]: fixed

File: settings.gradle (hunk 2 above)
Comment (bothejjms) [reply settings.gradle]: fixed

File: settings.gradle (hunk 3 above)
Comment (bothejjms) [reply settings.gradle]: fixed

File: src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java (hunk 4 above)
Comment (bothejjms) [reply src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java]: fixed

File: src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java (hunk 5 above)
Comment (bothejjms) [reply src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java]: rewrote it

File: src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java (hunk 6 above)
Comment (bothejjms) [reply src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java]: rewrote it

File: src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java (hunk 8 above)
Comment (bothejjms) [reply src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java]: rephrased

File: src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java (hunk 9 above)
Comment (bothejjms) [reply src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java]: changed it

File: src/main/java/com/ms/zookeeper/authz/ZkUserAuthorizer.java (hunk 10 above)
Comment (bothejjms) [reply src/main/java/com/ms/zookeeper/authz/ZkUserAuthorizer.java]: removed ZkUserAuthorizer for now

File: src/main/java/com/ms/zookeeper/authz/ZkUserAuthorizer.java (hunk 11 above)
Comment (bothejjms) [reply src/main/java/com/ms/zookeeper/authz/ZkUserAuthorizer.java]: removed ZkUserAuthorizer for now

File: src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java [hunk 17]
```diff
@@ -0,0 +1,114 @@
+package com.ms.zookeeper.auth;
```
Comment (andreikeis) [inline src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java:1]: I think we should keep original namespace - org.treadmill, not com.ms.

File: src/main/java/com/ms/zookeeper/authz/ZkFileAuthorizer.java [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+    }
+};
+
+public class ZkFileAuthorizer implements ZkAuthorizer, Runnable {
+
+    private static final Logger LOG = Logger.getLogger(ZkFileAuthorizer.class);
+
+    private static volatile ZkFileAuthorizer instance = null;
+
+    //expr and set of users who are allowed
```
Comment (andreikeis) [inline src/main/java/com/ms/zookeeper/authz/ZkFileAuthorizer.java]: what is proper style? I suspect // expr ... (with space after //. i am sure there is some sort of java style checker.

File: src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java (hunk 17 above)
Comment (bothejjms) [reply src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java:1]: I don't agree. This solution is not Treadmill specific why do we want to use a treadmill namespace? Treadmill is also not an organization.
For example the most well known external java lib is curator made by netflix and it is under com.netflix.curator

Comment (ebourgeois) [reply src/main/java/com/ms/zookeeper/auth/PluggableSASLAuthenticationProvider.java:1]: I agree with @bothejjms here, this is MS plugin, not Treadmill specific.

File: src/main/java/com/ms/zookeeper/authz/ZkFileAuthorizer.java (hunk 18 above)
Comment (ebourgeois) [reply src/main/java/com/ms/zookeeper/authz/ZkFileAuthorizer.java]: Yes, space after `//`; that is for single line comment, and `/** */` for multi line comment.

Comment (bothejjms) [reply src/main/java/com/ms/zookeeper/authz/ZkFileAuthorizer.java]: fixed

**Review by andreikeis — APPROVED** [review]
